frontend/components/instruments/SpatialInstrumentInternal.js

```diff
@@ -0,0 +1,180 @@
+   fixed by keeping everything playing at all times, just using 0 volume to "turn off" nodes.
+*/
+
+/*
+svgRender should be a component which takes 4 props (width: instrument width, height: instrument
+ height, x: current mouseX, y: current mouseY) and returns svg components to render (doesn't need
+ a wrapping svg tag, since this will be placed into the main svg tag of the instrument itself).
+ */
+const SpatialInstrumentInternal = (
+    {
+        soundPoints,
+        SvgRender=({width,height,x,y}) => <></>,
+        showPoints=true,
+        width=200,
+        height=200,
+        minPointRadius=2,
+        maxPointRadius=5,
+        updateSize=(w, h)=>{},
+    }
+) => {
+    const audioContextRef = useRef(new AudioContext());
+
+    const instrumentDiv = useRef(null);
+    const [mouseX, setMouseX] = useState(null);
+    const [mouseY, setMouseY] = useState(null);
+
+    const [instrumentWidth, setInstrumentWidth] = useState(null);
+    const [instrumentHeight, setInstrumentHeight] = useState(null);
+
+    useEffect(() => {
+        const {width: w, height: h} = instrumentDiv.current.getBoundingClientRect();
+        setInstrumentWidth(w);
+        setInstrumentHeight(h);
+    }, [width, height, instrumentDiv]);
+
+    useEffect(() => {
+        updateSize(instrumentWidth, instrumentHeight);
+    }, [instrumentWidth, instrumentHeight]);
+
+    // Instrument properties
+    const maxDist = 100;
+    const halfRange = 40;
+    const visualizeRange = true;
+
+    function sqDist(x1, y1, x2, y2) {
+        return (x1 - x2)*(x1 - x2) + (y1 - y2)*(y1 - y2);
+    }
+
+    function mouseActive() {
+        return mouseX !== null && mouseY !== null;
+    }
+
+    function sampleShouldPlay(point) {
+        if (!mouseActive()) {
+            return false;
+        }
+        return sqDist(point.x, point.y, mouseX, mouseY) <= maxDist*maxDist;
+    }
+
+    function sampleVolume(point) {
+        if (!mouseActive()) {
+            return 0;
+        }
+        if (!sampleShouldPlay(point)) return 0;
+        return 100/(sqDist(point.x, point.y, mouseX, mouseY)/(halfRange*halfRange) + 1);
+    }
+
+    function sampleRadius(point) {
+        const vol = sampleVolume(point);
+        return minPointRadius + (vol/100) * (maxPointRadius - minPointRadius);
+    }
+
+    function handleMouseLeave() {
+        setMouseX(null);
+        setMouseY(null);
+    }
+
+    function handleMouseMove(e) {
+        const rect = instrumentDiv.current.getBoundingClientRect();
+
+        // update width + height if out of sync
+        if (instrumentWidth !== rect.width) {
+            setInstrumentWidth(rect.width);
+        }
+        if (instrumentHeight !== rect.height) {
+            setInstrumentHeight(rect.height);
+        }
+
+        const x = e.clientX - rect.left; // x position within the element.
+        const y = e.clientY - rect.top;  // y position within the element.
+        setMouseX(x);
+        setMouseY(y);
+    }
+
+    return (
+        <>
```

Comment:
Since we're returning a single root element, we don't need to wrap in a fragment here.